File: examples/bun-app/src/handlers/payments.ts
/**
 * Payments namespace step handlers for refund processing.
 *
 * 4 sequential steps owned by the Payments team:
 *   ProcessRefundPayment -> UpdateLedger -> ReconcileAccount -> GenerateRefundReceipt
 *
 * Thin DSL wrappers that delegate to ../services/payments for business logic.
 */

import { defineHandler, PermanentError } from "@tasker-systems/tasker";
import * as svc from "../services/payments";
import {
	type PaymentsProcessGatewayResult,
	type PaymentsValidateEligibilityResult,
	ValidatePaymentEligibilityInputSchema,
} from "../services/schemas";

export const ProcessRefundPaymentHandler = defineHandler(
	"Payments.StepHandlers.ProcessRefundPaymentHandler",
	{
		inputs: {
			paymentId: "payment_id",
			refundAmount: "refund_amount",
		},
	},
	async ({ paymentId, refundAmount }) => {
		const parsed = ValidatePaymentEligibilityInputSchema.safeParse({
			paymentId,
			refundAmount,
		});

		if (!parsed.success) {
			const fields = parsed.error.issues
				.map((i) => i.path.join("."))
				.join(", ");
			throw new PermanentError(
				`Input validation failed: ${fields} — ${parsed.error.issues.map((i) => i.message).join("; ")}`,
			);
		}

		return svc.validateEligibility(parsed.data);
	},
);

export const UpdateLedgerHandler = defineHandler(
	"Payments.StepHandlers.UpdateLedgerHandler",
	{
		depends: { validationResult: "validate_payment_eligibility" },
	},
	async ({ validationResult }) =>
		svc.processGateway(
			validationResult as PaymentsValidateEligibilityResult | undefined,
		),
);

export const ReconcileAccountHandler = defineHandler(
	"Payments.StepHandlers.ReconcileAccountHandler",
	{
		depends: { refundResult: "process_gateway_refund" },
	},
	async ({ refundResult }) =>
		svc.updateRecords(refundResult as PaymentsProcessGatewayResult | undefined),
);

export const GenerateRefundReceiptHandler = defineHandler(
	"Payments.StepHandlers.GenerateRefundReceiptHandler",
	{
		depends: { refundResult: "process_gateway_refund" },
		inputs: { customerEmail: "customer_email" },
	},
	async ({ refundResult, customerEmail }) =>
		svc.notifyCustomer(
			refundResult as PaymentsProcessGatewayResult | undefined,
			customerEmail as string | undefined,
		),
);

File: examples/fastapi-app/app/handlers/payments.py
"""Payments namespace step handlers for refund processing.

4 sequential steps owned by the Payments team:
  ValidateEligibility -> ProcessGateway -> UpdateRecords -> NotifyCustomer

Thin DSL wrappers that delegate to app.services.payments for business logic.
"""

from __future__ import annotations

from tasker_core.step_handler.functional import depends_on, inputs, step_handler
from tasker_core.types import StepContext

from app.services import payments as svc
from app.services.types import (
    PaymentsProcessGatewayResult,
    PaymentsUpdateRecordsResult,
    PaymentsValidateEligibilityResult,
    ValidatePaymentEligibilityInput,
)


@step_handler("validate_eligibility")
@inputs(ValidatePaymentEligibilityInput)
def validate_eligibility(inputs: ValidatePaymentEligibilityInput, context: StepContext):
    # Input validation (required fields) is handled by the model's
    # @model_validator — see ValidatePaymentEligibilityInput in app/services/types.py.
    return svc.validate_eligibility(
        ValidatePaymentEligibilityInput(
            payment_id=inputs.payment_id or inputs.order_ref,
            refund_amount=inputs.refund_amount,
            refund_reason=inputs.refund_reason,
            partial_refund=inputs.partial_refund or False,
            order_ref=inputs.order_ref,
            customer_email=inputs.customer_email,
        )
    )


@step_handler("process_gateway")
@depends_on(eligibility=("validate_eligibility", PaymentsValidateEligibilityResult))
@inputs(ValidatePaymentEligibilityInput)
def process_gateway(
    eligibility: PaymentsValidateEligibilityResult,
    inputs: ValidatePaymentEligibilityInput,
    context: StepContext,
):
    return svc.process_gateway(
        eligibility=eligibility,
        refund_reason=inputs.refund_reason,
        partial_refund=inputs.partial_refund or False,
    )


@step_handler("update_records")
@depends_on(
    eligibility=("validate_eligibility", PaymentsValidateEligibilityResult),
    refund_result=("process_gateway", PaymentsProcessGatewayResult),
)
@inputs(ValidatePaymentEligibilityInput)
def update_records(
    eligibility: PaymentsValidateEligibilityResult,
    refund_result: PaymentsProcessGatewayResult,
    inputs: ValidatePaymentEligibilityInput,
    context: StepContext,
):
    return svc.update_records(
        eligibility=eligibility,
        refund_result=refund_result,
        refund_reason=inputs.refund_reason,
    )


@step_handler("notify_customer")
@depends_on(
    eligibility=("validate_eligibility", PaymentsValidateEligibilityResult),
    refund_result=("process_gateway", PaymentsProcessGatewayResult),
    records=("update_records", PaymentsUpdateRecordsResult),
)
@inputs(ValidatePaymentEligibilityInput)
def notify_customer(
    eligibility: PaymentsValidateEligibilityResult,
    refund_result: PaymentsProcessGatewayResult,
    records: PaymentsUpdateRecordsResult,
    inputs: ValidatePaymentEligibilityInput,
    context: StepContext,
):
    return svc.notify_customer(
        eligibility=eligibility,
        refund_result=refund_result,
        records=records,
        customer_email=inputs.customer_email,
    )
